### ISSUE TITLE
 Jump To Item Peak RMS/LUFS

### ISSUE THREAD

**Issue by JustinPerkinsMKE** [reaper-oss/sws#854]
There is a handy SWS action to jump to item peak sample (the loudest peak sample in the session) but what would be more useful to me is a way to jump to peak LUFS (or at least RMS) sections of an item using an action/script.

I would ask SWS but they seem to be non-responsive.

This goes along with request to normal items to a determined LUFS or RMS level.

This would be a HUGE benefit to mastering folks as well as others.

**More comments here:**
http://forum.cockos.com/showthread.php?t=183151

**Comment by nofishonfriday** [comment]
partly duplicate of #674.

**Comment by JustinPerkinsMKE** [comment]
Yes. This would be very useful. Often times, the sample peak level is not the loudness/most dense part of the song and the goal here is to move to the loudest section of the song.

Thanks for considering it.

**Comment by nofishonfriday** [comment]
With the added ReaScript loudness functions in v2.9.6 jumping to true peak / short term / momentary LUFS position and normalizing to LUFS is now possible via scripts. 

Example code here:
https://github.com/reaper-oss/sws/pull/882

**Comment by JustinPerkinsMKE** [comment]
Thanks @nofishonfriday I'm not the smartest when it comes to adding scripts etc. 

To be clear, I don't want to change the LUFS level with this action, that would already be done and locked in.

My goal would be to just select and item, and press a button to jump to the peak short term LUFS.

Is that possible?

**Comment by nofishonfriday** [comment]
I understood correctly, yes, possible to just jump to peak short term (or peak momentary) LUFS without changing LUFS level itself.

Below is a basic example Lua script.
When you select an audio item and run this it should put cursor to peak short term (should give same result as doing it via SWS/BR: Analyze loudness... - rightclick an entry - Go to maximum short term)

(You probably know that short term is a time _range_ of 3 seconds, so the cursor position indicates the beginning of this range. SWS/BR: Analyze loudness... automatically sets a time selection when jumping to peak short term to indicate this, but I'll leave this to the scripters).

If you plan more advanced stuff with this maybe ask for help in the ReaScript forum if necessary. :)

```
reaper.ClearConsole()

function msg(m)
  return reaper.ShowConsoleMsg(tostring(m) .. "\n")
end

function PrintResults()
  -- msg("process ok ?")
  -- msg(success)
   
  msg("short term max:")
  msg(shortTermMax)
  
  msg("short term max pos.:")
  msg( shortTermMaxPos)
  
end

item = reaper.GetSelectedMediaItem(0, 0)
if (item ~= nil) then
  take = reaper.GetActiveTake(item)
  if (take ~= nil) then
    
    -- reaper.NF_AnalyzeTakeLoudness2(), additionally gets shortTermMaxPos, momentaryMaxPos
    success = false
    analyzeTruePeak = false
    success, lufsIntegrated, range, truePeak, truePeakPos, shortTermMax, momentaryMax, shortTermMaxPos, momentaryMaxPos 
      = reaper.NF_AnalyzeTakeLoudness2(take, analyzeTruePeak)
    PrintResults()
    msg("\n")
    
    -- jump to shortTermMaxPos
    if (shortTermMaxPos >= 0) then
      reaper.PreventUIRefresh(1)
      reaper.SetEditCurPos2(NULL, shortTermMaxPos, true, false)
      reaper.PreventUIRefresh(-1)
    end
     
  end -- if (take ~= nil)
end -- if (item ~= nil)
```

**Comment by JustinPerkinsMKE** [comment]
@nofishonfriday Thanks, this seems to be working great. I can probably pair it with a command to start playback as well and make a custom script that finds the loudest short-term LUFS and automatically starts playing.

One last thing here.

Is there a way to prevent the Rescript Console Output box from appearing each time I do this? I don't need or want to see that, just to have the cursor move and start playback.

Thanks.

**Comment by nofishonfriday** [comment]
Ah yes, sorry, console output was just for demonstartion purpose.
Here's without console output:

```
-- reaper.ClearConsole()

function msg(m)
  return reaper.ShowConsoleMsg(tostring(m) .. "\n")
end

function PrintResults()
  -- msg("process ok ?")
  -- msg(success)
   
  -- msg("short term max:")
  -- msg(shortTermMax)
  
  -- msg("short term max pos.:")
  -- msg( shortTermMaxPos)
  
end

item = reaper.GetSelectedMediaItem(0, 0)
if (item ~= nil) then
  take = reaper.GetActiveTake(item)
  if (take ~= nil) then
    
    -- reaper.NF_AnalyzeTakeLoudness2(), additionally gets shortTermMaxPos, momentaryMaxPos
    success = false
    analyzeTruePeak = false
    success, lufsIntegrated, range, truePeak, truePeakPos, shortTermMax, momentaryMax, shortTermMaxPos, momentaryMaxPos 
      = reaper.NF_AnalyzeTakeLoudness2(take, analyzeTruePeak)
    -- PrintResults()
    -- msg("\n")
    
    -- jump to shortTermMaxPos
    if (shortTermMaxPos >= 0) then
      reaper.PreventUIRefresh(1)
      reaper.SetEditCurPos2(NULL, shortTermMaxPos, true, false)
      reaper.PreventUIRefresh(-1)
    end
     
  end -- if (take ~= nil)
end -- if (item ~= nil)
```

**Comment by JustinPerkinsMKE** [comment]
@nofishonfriday This is great, thanks. My only question now is if there is a way to pre-analyze items so it can play more quickly or will it always need to do that scan?

It's not too slow and maybe when I move to the iMac Pro eventually it'll be even faster.

**Comment by nofishonfriday** [comment]
Well, it needs to make the analyzing once at least (I mostly just exposed how SWS/BR: Analyze loudness... does it to ReaScript to be honest) but with a suitable script e.g. a workflow like this should be possible I think:

Batch analyze loudness of all selected items first (only once), then use that data for jumping to max short term pos. of selected item and start play. 

But again, that's something I'll leave to the scripters.

**Comment by JustinPerkinsMKE** [comment]
@nofishonfriday I think it's safe to call this one done. Please let me know how to donate.

I may have another similar request though. I also use WaveLab software and this is where I partially got the idea for this action. Typically speaking, Maximum Short-Term LUFS is a good way to get the loudest parts of a group of items to be the same.

WaveLab has another option called "Top Of Loudness Range" which normalizes all items/clips so that the top of their loudness range is the same.

I was working in WaveLab last night on a compilation of live tracks from various artists and recording qualities. Using  "Top Of Loudness Range" sounded more natural to the ear than "Maximum Short-Term LUFS".

I'm wondering if it would be possible to make an alternate script that does that as well?

I asked the developer of WaveLab what  "Top Of Loudness Range" really means and he said this:

**"Loudness Range" refers to the R128 defintion of "Loudness Range".
This is before all a mathematical formula, but to make it more clear: the loudness varies between an average minimum (which excludes silences), a maximum one (which excludes short peaks).
What you refer to, is this last value.**

**Comment by nofishonfriday** [comment]
Happy to hear it's working. 
My Paypal adress is: o.keitel@web.de

As to the new one:
We can get loudness range also via scripts so assume it should be possible.

I'm currently still having a little trouble understanding how this actually works though.

- Is there anything written about it in the Wavelab manual you can forward to me ?
- Do you specify a target value when normalizing to "Top of loudness range" ?

I'll try to do an example, say:

Specified target level (if there's one to specify): -10 LUFS

Item A:
Integrated loudness: -20LUFS Range: 10 LU
-> avrg, min.: -25 LUFS, avrg. max ( = top of loudness range).; -15 LUFS
-> raise this item 5LU to match the specified target (-15 LUFS + 5LU = -10LUFS).

Does this sound right ?

**Comment by JustinPerkinsMKE** [comment]
@nofishonfriday I'll try to see if there is anything more in the manual. You do specify a target just like the Max Short Term LUFS but it seems to act differently, and perhaps be more natural to the human ear. I was really impressed how well it handled a mixed bag of live tracks from different artists, some with full band and some just acoustic.

Maybe I could process a few files with both settings for you to analyze?

I honestly don't know the details real well.

**Comment by JustinPerkinsMKE** [comment]
@nofishonfriday Let me know about the donation. I'm not sure how long these things take or how much your time is worth.

Thank you again.

**Comment by nofishonfriday** [comment]
Donation received, thank you.

Yeah, a few files (two versions of each: unprocessed - normalized with "Top of loudness" option) would help.

**Comment by JustinPerkinsMKE** [comment]
@nofishonfriday Will do. I'll try to find some material I can send you with a few normalization options for comparison.

It may take me a couple days to get it together.

**Comment by nofishonfriday** [comment]
Ok, no hurry. :)

Maybe in the meantime I'll try to put together a script with what I think could be the right approach for you to test / comapre.

**Comment by JustinPerkinsMKE** [comment]
@nofishonfriday Thanks!

**Comment by JustinPerkinsMKE** [comment]
@nofishonfriday Also, the new script could share the same target LUFS window as the script you just made. It would just intemperate the level differently.

**Comment by JustinPerkinsMKE** [comment]
@nofishonfriday To help respect the privacy of the music I will email you a download link to your PayPal ID address.

The zip file will have 3 subfolders:

Files normalized to -18LUFS Short Term MAX
Files normalized to -18LUFS Top Loundess Range
Original files (WAV and WaveLab montage files which you probably don't need)

Let me know what you find about them or if the difference is worth chasing.

I tried to pick a variety of music to help demonstrate.

**Comment by nofishonfriday** [comment]
Got the files, thanks.
I'll have a look and get back to you.

**Comment by JustinPerkinsMKE** [comment]
@nofishonfriday Thanks!. Sorry for the goofy music, I just grabbed a few things I had recently worked on for others. If you need anything else, let me know.

**Comment by nofishonfriday** [comment]
_Also, the new script could share the same target LUFS window as the script you just made. It would just intemperate the level differently._

That's the plan. :)

**Comment by JustinPerkinsMKE** [comment]
Thanks! I meant "interpret" and not "intemperate" but I guess you knew what I meant.

**Comment by nofishonfriday** [comment]
Question about the files you sent me:

Did you batch normalize these to -18LUFS top of loudness range, or did you do it one by one ? (I assume there's a batch normalizing feature in WL ?)

What I'm trying to find out, is "top of loudness range" an overall value, calculated from all to be processed files or is it calculated per file. (Btw. did you check already if there's anything useful about it in the manual ?)

Do you know by chance ?
If not, could you please do the following:

Take two of the unprocessed files
- normalize to top of loudness one by one
- batch normalize to top of loudness 

and send me the respective resulting files.

Thanks.

**Comment by JustinPerkinsMKE** [comment]
WaveLab does this one by one. Each clip (or item in REAPER terms) is processed independently.

This is how the current LUFS version you have made works right? It seems to be. The goal is to get all the items to be the same in level. Not a global normalize, but an item by item normalize.

The files I've sent are one by one, I would have to see if there is an easy way to it as an overall thing but that is not really the end goal anyway.

Here are some things in the manual about it:

This pop-up menu allows you to select a reference: the loudness of the entire file (EBU R-128 recommendation), the average loudest 3 second audio section (Top of Loudness Range), or the loudest 3 seconds audio section (Maximum Short-Term Loudness).

Top of a loudness range, that is, the average loudest 3 second audio section of a file. This ensures that a single unusually loud sound is not taken into account for the reference.
• Maximum short-term loudness, that is, the maximum loudness that is found in a 3 second audio section of the file, for example, the loudness of a short music passage.

**Comment by nofishonfriday** [comment]
This info and the excerpt from the manual is helpful, thanks.

**Comment by nofishonfriday** [comment]
Closing as a solution via script is now available (sse above example script).
